newIDE/app/src/LayersList/index.js

```diff
@@ -20,96 +19,149 @@ import Background from '../UI/Background';
+  }>({});
+  const draggedLayerIndexRef = React.useRef<number | null>(null);
 
-  _onLayerModified = () => {
-    if (this.props.unsavedChanges)
-      this.props.unsavedChanges.triggerUnsavedChanges();
-    this.forceUpdate();
+  const {
+    layersContainer,
+    onEditEffects,
+    onEdit,
+    width,
+    onRenameLayer,
+    onRemoveLayer,
+    unsavedChanges,
+  } = props;
+
+  const onLayerModified = () => {
+    if (unsavedChanges) unsavedChanges.triggerUnsavedChanges();
+    forceUpdate();
   };
 
-  render() {
-    const { layersContainer, onEditEffects, onEdit, width } = this.props;
-
-    const layersCount = layersContainer.getLayersCount();
-    const containerLayersList = mapReverseFor(0, layersCount, i => {
-      const layer = layersContainer.getLayerAt(i);
-      const layerName = layer.getName();
-      const isLightingLayer = layer.isLightingLayer();
-
-      return (
-        <SortableLayerRow
-          index={layersCount - 1 - i}
-          key={'layer-' + layerName}
-          layer={layer}
-          layerName={layerName}
-          isLightingLayer={isLightingLayer}
-          nameError={this.state.nameErrors[layerName]}
-          effectsCount={layer.getEffects().getEffectsCount()}
-          onEditEffects={() => onEditEffects(layer)}
-          onEdit={() => onEdit(layer)}
-          onBlur={event => {
-            const newName = event.target.value;
-            if (layerName === newName) return;
-
-            let success = true;
-            if (layersContainer.hasLayerNamed(newName)) {
-              success = false;
-            } else {
-              this.props.onRenameLayer(layerName, newName, doRename => {
-                if (doRename)
-                  layersContainer.getLayer(layerName).setName(newName);
-              });
-            }
-
-            this.setState({
-              nameErrors: {
-                ...this.state.nameErrors,
-                [layerName]: !success,
-              },
-            });
-          }}
-          onRemove={() => {
-            this.props.onRemoveLayer(layerName, doRemove => {
-              if (!doRemove) return;
+  const onDropLayer = (targetIndex: number) => {
+    const { current: draggedLayerIndex } = draggedLayerIndexRef;
+    if (draggedLayerIndex === null) return;
 
-              layersContainer.removeLayer(layerName);
-              this._onLayerModified();
-            });
-          }}
-          isVisible={layer.getVisibility()}
-          onChangeVisibility={visible => {
-            layer.setVisibility(visible);
-            this._onLayerModified();
-          }}
-          width={width}
-        />
+    if (targetIndex !== draggedLayerIndex) {
+      layersContainer.moveLayer(
+        draggedLayerIndex,
+        targetIndex < draggedLayerIndex ? targetIndex + 1 : targetIndex
       );
-    });
+      onLayerModified();
+    }
+    draggedLayerIndexRef.current = null;
+  };
+
+  const DropTarget = React.useMemo(() => makeDropTarget('layers-list'), []);
```

Comment:
I think this would work even outside of the component

Reply:
True, I'll change it